### PR TITLE
streams: pipeTo a WS with HWM>1

### PR DESCRIPTION
**Pull request by isonmad**
`pipeTo` in the reference implementation currently effectively ignores the WritableStream's highWaterMark. Or is that intended behavior?

### CODE REVIEW

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/216519166)

# Firefox (nightly channel) #
Testing web-platform-tests at revision 05856f33f88f853fa4fc859e333e0cb1e690ca5d
Using browser at version BuildID 20170329142008; SourceStamp 6ea713ccc9abea93126423fefb855d0e051c95e2
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>4 tests ran</summary>
<details>
<summary><a href="http://w3c-test.org/submissions/5262/streams/piping/flow-control.dedicatedworker.html">/streams/piping/flow-control.dedicatedworker.html</a></summary>


|                                                                                        Subtest                                                                                         | Results |                Messages               |
|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|---------|---------------------------------------|
|                                                                                                                                                                                        | OK      |                                       |
| `Piping from a non-empty ReadableStream into a WritableStream that does not desire chunks`                                                                                             | FAIL    | `ReadableStream is not defined`       |
| `Piping from a non-empty ReadableStream into a WritableStream that does not desire chunks, but then does`                                                                              | FAIL    | `ReadableStream is not defined`       |
| `Piping from an empty ReadableStream into a WritableStream that does not desire chunks, but then the readable stream becomes non-empty and the writable stream starts desiring chunks` | FAIL    | `ReadableStream is not defined`       |
| `Piping from a ReadableStream to a WritableStream that desires more chunks before finishing with previous ones`                                                                        | FAIL    | `CountQueuingStrategy is not defined` |
| `Piping to a WritableStream that does not consume the writes fast enough exerts backpressure on the ReadableStream`                                                                    | FAIL    | `ReadableStream is not defined`       |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5262/streams/piping/flow-control.html">/streams/piping/flow-control.html</a></summary>


|                                                                                        Subtest                                                                                         | Results |                Messages               |
|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|---------|---------------------------------------|
|                                                                                                                                                                                        | OK      |                                       |
| `Piping from a non-empty ReadableStream into a WritableStream that does not desire chunks`                                                                                             | FAIL    | `ReadableStream is not defined`       |
| `Piping from a non-empty ReadableStream into a WritableStream that does not desire chunks, but then does`                                                                              | FAIL    | `ReadableStream is not defined`       |
| `Piping from an empty ReadableStream into a WritableStream that does not desire chunks, but then the readable stream becomes non-empty and the writable stream starts desiring chunks` | FAIL    | `ReadableStream is not defined`       |
| `Piping from a ReadableStream to a WritableStream that desires more chunks before finishing with previous ones`                                                                        | FAIL    | `CountQueuingStrategy is not defined` |
| `Piping to a WritableStream that does not consume the writes fast enough exerts backpressure on the ReadableStream`                                                                    | FAIL    | `ReadableStream is not defined`       |

</details>
<details>
<summary><a href="https://w3c-test.org/submissions/5262/streams/piping/flow-control.serviceworker.https.html">/streams/piping/flow-control.serviceworker.https.html</a></summary>


|                                                                                        Subtest                                                                                         | Results |                Messages               |
|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|---------|---------------------------------------|
|                                                                                                                                                                                        | OK      |                                       |
| `Service worker test setup`                                                                                                                                                            | PASS    |                                       |
| `Piping from a non-empty ReadableStream into a WritableStream that does not desire chunks`                                                                                             | FAIL    | `ReadableStream is not defined`       |
| `Piping from a non-empty ReadableStream into a WritableStream that does not desire chunks, but then does`                                                                              | FAIL    | `ReadableStream is not defined`       |
| `Piping from an empty ReadableStream into a WritableStream that does not desire chunks, but then the readable stream becomes non-empty and the writable stream starts desiring chunks` | FAIL    | `ReadableStream is not defined`       |
| `Piping from a ReadableStream to a WritableStream that desires more chunks before finishing with previous ones`                                                                        | FAIL    | `CountQueuingStrategy is not defined` |
| `Piping to a WritableStream that does not consume the writes fast enough exerts backpressure on the ReadableStream`                                                                    | FAIL    | `ReadableStream is not defined`       |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5262/streams/piping/flow-control.sharedworker.html">/streams/piping/flow-control.sharedworker.html</a></summary>


|                                                                                        Subtest                                                                                         | Results |                Messages               |
|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|---------|---------------------------------------|
|                                                                                                                                                                                        | OK      |                                       |
| `Piping from a non-empty ReadableStream into a WritableStream that does not desire chunks`                                                                                             | FAIL    | `ReadableStream is not defined`       |
| `Piping from a non-empty ReadableStream into a WritableStream that does not desire chunks, but then does`                                                                              | FAIL    | `ReadableStream is not defined`       |
| `Piping from an empty ReadableStream into a WritableStream that does not desire chunks, but then the readable stream becomes non-empty and the writable stream starts desiring chunks` | FAIL    | `ReadableStream is not defined`       |
| `Piping from a ReadableStream to a WritableStream that desires more chunks before finishing with previous ones`                                                                        | FAIL    | `CountQueuingStrategy is not defined` |
| `Piping to a WritableStream that does not consume the writes fast enough exerts backpressure on the ReadableStream`                                                                    | FAIL    | `ReadableStream is not defined`       |

</details>
</details>

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/216519167)

# Chrome (unstable channel) #
Testing web-platform-tests at revision 05856f33f88f853fa4fc859e333e0cb1e690ca5d
Using browser at version 59.0.3053.3 dev
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>4 tests ran</summary>
<details>
<summary><a href="http://w3c-test.org/submissions/5262/streams/piping/flow-control.dedicatedworker.html">/streams/piping/flow-control.dedicatedworker.html</a></summary>


|                                                                                        Subtest                                                                                         | Results |             Messages            |
|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|---------|---------------------------------|
|                                                                                                                                                                                        | OK      |                                 |
| `Piping from a non-empty ReadableStream into a WritableStream that does not desire chunks`                                                                                             | FAIL    | `WritableStream is not defined` |
| `Piping from a non-empty ReadableStream into a WritableStream that does not desire chunks, but then does`                                                                              | FAIL    | `WritableStream is not defined` |
| `Piping from an empty ReadableStream into a WritableStream that does not desire chunks, but then the readable stream becomes non-empty and the writable stream starts desiring chunks` | FAIL    | `WritableStream is not defined` |
| `Piping from a ReadableStream to a WritableStream that desires more chunks before finishing with previous ones`                                                                        | FAIL    | `WritableStream is not defined` |
| `Piping to a WritableStream that does not consume the writes fast enough exerts backpressure on the ReadableStream`                                                                    | FAIL    | `WritableStream is not defined` |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5262/streams/piping/flow-control.html">/streams/piping/flow-control.html</a></summary>


|                                                                                        Subtest                                                                                         | Results |             Messages            |
|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|---------|---------------------------------|
|                                                                                                                                                                                        | OK      |                                 |
| `Piping from a non-empty ReadableStream into a WritableStream that does not desire chunks`                                                                                             | FAIL    | `WritableStream is not defined` |
| `Piping from a non-empty ReadableStream into a WritableStream that does not desire chunks, but then does`                                                                              | FAIL    | `WritableStream is not defined` |
| `Piping from an empty ReadableStream into a WritableStream that does not desire chunks, but then the readable stream becomes non-empty and the writable stream starts desiring chunks` | FAIL    | `WritableStream is not defined` |
| `Piping from a ReadableStream to a WritableStream that desires more chunks before finishing with previous ones`                                                                        | FAIL    | `WritableStream is not defined` |
| `Piping to a WritableStream that does not consume the writes fast enough exerts backpressure on the ReadableStream`                                                                    | FAIL    | `WritableStream is not defined` |

</details>
<details>
<summary><a href="https://w3c-test.org/submissions/5262/streams/piping/flow-control.serviceworker.https.html">/streams/piping/flow-control.serviceworker.https.html</a></summary>


|                                                                                        Subtest                                                                                         | Results |             Messages            |
|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|---------|---------------------------------|
|                                                                                                                                                                                        | OK      |                                 |
| `Service worker test setup`                                                                                                                                                            | PASS    |                                 |
| `Piping from a non-empty ReadableStream into a WritableStream that does not desire chunks`                                                                                             | FAIL    | `WritableStream is not defined` |
| `Piping from a non-empty ReadableStream into a WritableStream that does not desire chunks, but then does`                                                                              | FAIL    | `WritableStream is not defined` |
| `Piping from an empty ReadableStream into a WritableStream that does not desire chunks, but then the readable stream becomes non-empty and the writable stream starts desiring chunks` | FAIL    | `WritableStream is not defined` |
| `Piping from a ReadableStream to a WritableStream that desires more chunks before finishing with previous ones`                                                                        | FAIL    | `WritableStream is not defined` |
| `Piping to a WritableStream that does not consume the writes fast enough exerts backpressure on the ReadableStream`                                                                    | FAIL    | `WritableStream is not defined` |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5262/streams/piping/flow-control.sharedworker.html">/streams/piping/flow-control.sharedworker.html</a></summary>


|                                                                                        Subtest                                                                                         | Results |             Messages            |
|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|---------|---------------------------------|
|                                                                                                                                                                                        | OK      |                                 |
| `Piping from a non-empty ReadableStream into a WritableStream that does not desire chunks`                                                                                             | FAIL    | `WritableStream is not defined` |
| `Piping from a non-empty ReadableStream into a WritableStream that does not desire chunks, but then does`                                                                              | FAIL    | `WritableStream is not defined` |
| `Piping from an empty ReadableStream into a WritableStream that does not desire chunks, but then the readable stream becomes non-empty and the writable stream starts desiring chunks` | FAIL    | `WritableStream is not defined` |
| `Piping from a ReadableStream to a WritableStream that desires more chunks before finishing with previous ones`                                                                        | FAIL    | `WritableStream is not defined` |
| `Piping to a WritableStream that does not consume the writes fast enough exerts backpressure on the ReadableStream`                                                                    | FAIL    | `WritableStream is not defined` |

</details>
</details>

**Comment by domenic** [comment]
The corresponding spec bug is https://github.com/whatwg/streams/issues/653. @ricea has been meaning to work on that I believe, and we should indeed fix the reference implementation.

File: streams/piping/flow-control.js [hunk 1]
```diff
@@ -125,6 +125,51 @@ promise_test(() => {
    'stream becomes non-empty and the writable stream starts desiring chunks');
 
 promise_test(() => {
+  const unreadChunks = ['b', 'c', 'd'];
+
+  const rs = recordingReadableStream({
+    pull(controller) {
+      controller.enqueue(unreadChunks.shift());
+      if (unreadChunks.length === 0)
```
Comment (ricea) [inline streams/piping/flow-control.js]: Style nit: In stream code we always use { } with if, even single-line.

File: streams/piping/flow-control.js [hunk 2]
```diff
@@ -125,6 +125,51 @@ promise_test(() => {
...
+          resolveWritePromise = resolve;
+        });
+      }
+      return undefined;
+    }
+  }, new CountQueuingStrategy({ highWaterMark: 3 }));
+
+  const writer = ws.getWriter();
+  const firstWritePromise = writer.write('a');
+  assert_equals(writer.desiredSize, 2, 'after writing the writer\'s desiredSize must be 3');
```
Comment (ricea) [inline streams/piping/flow-control.js]: "must be 2"

**Review by ricea — DISMISSED** [review]
As the standard is currently written, implementations are not required to pass this test.

I'm planning to make implementations use the queue, but even after the change implementations will be able to take a "reasonable" amount of time to fill the queue, in order to provide freedom to optimise chunk size as implied by

> * Otherwise, WritableStreamDefaultWriterGetDesiredSize(writer) may be used to determine the flow rate heuristically, e.g. by delaying reads while it is judged to be "low" compared to the size of chunks that have been typically read.

If you are able to propose better standard language at https://github.com/whatwg/streams/issues/653 that would be really great.

**Comment by w3c-bots** [comment]
These tests are now available on [w3c-test.org](https://w3c-test.org/submissions/5262/)

**Comment by domenic** [comment]
I addressed the code review comments, so if @ricea is happy with this (and the spec change) we can merge them all together.

**Comment by ricea** [comment]
lgtm

**Pull request merged**